ecosystem/python/sdk/aptos_sdk/secp256k1_ecdsa.py

```diff
@@ -0,0 +1,233 @@
+# Copyright © Aptos Foundation
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+import hashlib
+import unittest
+from typing import cast
+
+from ecdsa import SECP256k1, SigningKey, VerifyingKey, util
+
+from . import asymmetric_crypto
+from .bcs import Deserializer, Serializer
+
+
+class PrivateKey(asymmetric_crypto.PrivateKey):
+    LENGTH: int = 32
+
+    key: SigningKey
+
+    def __init__(self, key: SigningKey):
+        self.key = key
+
+    def __eq__(self, other: object):
+        if not isinstance(other, PrivateKey):
+            return NotImplemented
+        return self.key == other.key
+
+    def __str__(self):
+        return self.hex()
+
+    @staticmethod
+    def from_str(value: str) -> PrivateKey:
+        if value[0:2] == "0x":
+            value = value[2:]
+        if len(value) != PrivateKey.LENGTH * 2:
+            raise Exception("Length mismatch")
+        return PrivateKey(
+            SigningKey.from_string(bytes.fromhex(value), SECP256k1, hashlib.sha3_256)
+        )
+
+    def hex(self) -> str:
+        return f"0x{self.key.to_string().hex()}"
+
+    def public_key(self) -> PublicKey:
+        return PublicKey(self.key.verifying_key)
+
+    @staticmethod
+    def random() -> PrivateKey:
+        return PrivateKey(
+            SigningKey.generate(curve=SECP256k1, hashfunc=hashlib.sha3_256)
+        )
+
+    def sign(self, data: bytes) -> Signature:
+        sig = self.key.sign_deterministic(data, hashfunc=hashlib.sha3_256)
+        n = SECP256k1.generator.order()
+        r, s = util.sigdecode_string(sig, n)
+        # The signature is valid for both s and -s, normalization ensures that only s < n // 2 is valid
+        if s > (n // 2):
+            mod_s = (s * -1) % n
+            sig = util.sigencode_string(r, mod_s, n)
+        return Signature(sig)
+
+    @staticmethod
+    def deserialize(deserializer: Deserializer) -> PrivateKey:
+        key = deserializer.to_bytes()
+        if len(key) != PrivateKey.LENGTH:
+            raise Exception("Length mismatch")
+
+        return PrivateKey(SigningKey.from_string(key, SECP256k1, hashlib.sha3_256))
+
+    def serialize(self, serializer: Serializer):
+        serializer.to_bytes(self.key.to_string())
+
+
+class PublicKey(asymmetric_crypto.PublicKey):
+    LENGTH: int = 64
+    PREFIX_LENGTH: int = 65
+
+    key: VerifyingKey
+
+    def __init__(self, key: VerifyingKey):
+        self.key = key
+
+    def __eq__(self, other: object):
+        if not isinstance(other, PublicKey):
+            return NotImplemented
+        return self.key == other.key
+
+    def __str__(self) -> str:
+        return self.hex()
+
+    @staticmethod
+    def from_str(value: str) -> PublicKey:
+        if value[0:2] == "0x":
+            value = value[2:]
+        if (
+            len(value) != PublicKey.LENGTH * 2
+            and len(value) != PublicKey.PREFIX_LENGTH * 2
+        ):
```

Comment:
Should probably make these a more common parsing function since you're always checking length.

Though, why prefix length and length?  You can provide both lengths?  Shouldn't this be:

```
len(value) != (PublicKey.PREFIX_LENGTH + PublicKey.LENGTH) * 2
```

Reply:
ahh... LENGTH_WITH_PREFIX is probably a better phrasing